### ISSUE TITLE
wishlist: hooks

### ISSUE THREAD

**Issue by tianon** [infosiftr/hocker#5]
It'd be great to have some kind of support for "hooks" -- bits of shell that get executed at different times during a `hocker_run`.

For example, if I have a website running behind an `nginx` instance, then when my website container actually does a restart (for example if I'm using smart-restarting), I want to `docker kill -s HUP nginx` so that it picks up the new IP address, but I don't want to run that every time I invoke the hocker script (since it's potentially somewhat expensive).

For example:

```bash
hocker_run --hook-post-start 'docker kill -s HUP nginx || :' wordpress
```

Or something like:

```bash
hocker_run --hook-pre-stop '# some complex means of telling the application to stop cleanly' some-app
```

**Comment by tianon** [comment]
Oh!  A perfect example for `--hook-pre-stop` is Jenkins!  It has a function whereby Jenkins can be put in "preparing for shutdown" mode, which would be a good thing for this hook to do before the `SIGTERM` is sent.

**Comment by tianon** [comment]
Another fun example would be using `--hook-pre-stop` to force stopping to fail if the application isn't safe to shut down right now.  For example, if we've got a MongoDB cluster, it'd be useful to have it step down before we stop the node, but if it fails to step down, making the stop fail completely would be a sane thing to do. :metal:

**Comment by yosifkit** [comment]
The quoted bash code to pass to the hooks would be annoying but that should be solvable by passing a function instead (as a string).  If we advocate that the user only passes a function then we can have hocker pass a specific set of arguments to it; not sure if it needs any, but it could be nice.

**Comment by tianon** [comment]
Yeah, I was thinking about arguments, but then I realized we'll have the same `$scriptDir` etc that we have normally, and we can even add extra variables before we invoke these things (such as `$name`, etc) -- some of those they'd get naturally just by the nature of where in the script they get `eval`'d, but it'd probably be good to document or "bless" some of those as explicitly "safe to use in hooks".

**Comment by tianon** [comment]
Re-reading this thread all these years later, I don't like it -- hooks that happen during "stop" but _only_ if `hocker` is the thing that did the stopping are exactly the kind of thing I dislike about Compose.  The idea about _conditionals_ for stopping that only get invoked if we're actually stopping are interesting, but the rest make me moderately uncomfortable. 😅

(This is rare!  Usually I agree with past Tianon, but he's out past his skis here.)
